### ISSUE TITLE
Enable hardware-simulated-keyDownEvent in production (currently building for windows)

### ISSUE THREAD

**Issue by Eerey** [flutter/flutter#98742]
## Use case

I want to be able to simulate keyDownEvents in production like in https://api.flutter.dev/flutter/flutter_test/simulateKeyDownEvent.html

Requirements: **There is no hardware-keyboard available (!)**, but we're simulating an on-screen keyboard (and want to style it here and there differently). When tapping on the on-screen keyboard, i want to simulate hardware keystrokes.

![image](https://user-images.githubusercontent.com/7249645/154674386-6c06d3cb-b8c1-420d-89fa-cea387997db9.png)

## Proposal

Provide a function similar to e.g. `KeyEventSimulator.simulateKeyDownEvent(LogicalKeyboardKey.keyA);`

**Comment by darshankawar** [comment]
@Eerey 
Can you please elaborate when you say enable in production ?
Does it mean when you try that in debug mode ? Please check [this](https://github.com/flutter/flutter/issues/96022) similar issue and underlying comments and see if they help to achieve your use case.

**Comment by Eerey** [comment]
It does neither work on production nor in debug. Actually, I'd like to have it enabled on any mode.

**Comment by darshankawar** [comment]
> Please check [this](https://github.com/flutter/flutter/issues/96022) similar issue and underlying comments and see if they help to achieve your use case.

Did you get a chance to go through this linked issue to see if any of the underlying comments in it makes sense for your case ?

**Comment by Eerey** [comment]
@darshankawar Yes I took a look into it, but it is tied to `Testing` and i want to use it in Production aswell (mode-independent).

It seems i can not use any of the suggested comments on that related issue.

My issue remains: How do I simulate hardware keyDown/Up Events mode-independently (debug/profile/release)?

**Comment by Eerey** [comment]
Related: https://github.com/flutter/flutter/pull/76072

**Comment by darshankawar** [comment]
Thank for further details.

**Comment by gspencergoog** [comment]
Correct, it's not going to work to just call `RawKeyboard.handleRawKeyEvent`, since the actual text input is a cooperation between the engine and the framework, and you're just simulating the framework side when you do that. Because input method editors need to provide an edit stream rather than key events, we need to get the text input through the engine.

For your use case, have you thought of just calling the relevant Windows APIs to simulate key events, using Dart's FFI? The [win32](https://pub.dev/packages/win32) package can help a lot with this. Then all of the correct things will happen with the events.

cc @dkwingsmt

**Comment by Eerey** [comment]
@gspencergoog Thanks for your response. I like the idea and will definitely try out the [win32 package](https://pub.dev/packages/win32).

However, I think it would really complete the flutter input-mechanics if a full-fletched simulation of hardware-inputs are an option.

Why? Because my next question would be: How do I do it on MacOS / Linux? It will be the same question and would require more `ffi` etc. I don't have anything against `ffi`, but it bites me sometimes and is quite a hassle when building with a single code base. Actually, if you start to manage certain platform-specific libraries for Windows/MacOS/Linux - what is the point of having a single code base? My paradigm and understanding of flutter is: Write code once, run everywhere. Implementing a virtual-keyboard-overlay for each platform does not follow that mindset.

I would really appreciate the option to work on an `On-Screen Keyboard` for every platform without worrying of any c-compiled libraries.

Would it be technically possible?

**Comment by gspencergoog** [comment]
Yes, it's technically possible.

And it would be nice, I agree, and would actually help our testing infrastructure too, but our resources are limited, and this is not a use case that is often requested, so other tasks would take priority for some time before we got to this.

**Comment by Eerey** [comment]
@gspencergoog Cool, I don't want to rush anything, but I'm **VERY glad** to hear that. Thanks for your response!

Keep on going, still loving the product.

**Comment by dkwingsmt** [comment]
I agree that it's important for many applications to forge events. But make it work exactly like native takes more than just framework. Each OS has its own range of support for different types of events, especially Win32 has an extremely complicated protocol. 

But in the end, all keyboard effects come down to key events and text input. Maybe we should just provide a way to forge text input changes.
